libs/langchain/langchain/vectorstores/neo4j_vector.py

```diff
@@ -0,0 +1,623 @@
+        """Run more texts through the embeddings and add to the vectorstore.
+
+        Args:
+            texts: Iterable of strings to add to the vectorstore.
+            metadatas: Optional list of metadatas associated with the texts.
+            kwargs: vectorstore specific parameters
+
+        Returns:
+            List of ids from adding the texts into the vectorstore.
+        """
+        embeddings = self.embedding_function.embed_documents(list(texts))
+        return self.add_embeddings(
+            texts=texts, embeddings=embeddings, metadatas=metadatas, ids=ids, **kwargs
+        )
+
+    def similarity_search(
+        self,
+        query: str,
+        k: int = 4,
+        **kwargs: Any,
+    ) -> List[Document]:
+        """Run similarity search with Neo4jVector.
+
+        Args:
+            query (str): Query text to search for.
+            k (int): Number of results to return. Defaults to 4.
+
+        Returns:
+            List of Documents most similar to the query.
+        """
+        embedding = self.embedding_function.embed_query(text=query)
+        return self.similarity_search_by_vector(
+            embedding=embedding,
+            k=k,
+        )
+
+    def similarity_search_with_score(
+        self, query: str, k: int = 4
+    ) -> List[Tuple[Document, float]]:
+        """Return docs most similar to query.
+
+        Args:
+            query: Text to look up documents similar to.
+            k: Number of Documents to return. Defaults to 4.
+
+        Returns:
+            List of Documents most similar to the query and score for each
+        """
+        embedding = self.embedding_function.embed_query(query)
+        docs = self.similarity_search_with_score_by_vector(embedding=embedding, k=k)
+        return docs
+
+    def similarity_search_with_score_by_vector(
+        self, embedding: List[float], k: int = 4
+    ) -> List[Tuple[Document, float]]:
+        """
+        Perform a similarity search in the Neo4j database using a
+        given vector and return the top k similar documents with their scores.
+
+        This method uses a Cypher query to find the top k documents that
+        are most similar to a given embedding. The similarity is measured
+        using a vector index in the Neo4j database. The results are returned
+        as a list of tuples, each containing a Document object and
+        its similarity score.
+
+        Args:
+            embedding (List[float]): The embedding vector to compare against.
+            k (int, optional): The number of top similar documents to retrieve.
+
+        Returns:
+            List[Tuple[Document, float]]: A list of tuples, each containing
+                                a Document object and its similarity score.
+        """
+
+        read_query = (
+            "CALL db.index.vector.queryNodes($index, $k, $embedding) "
+            "YIELD node, score "
+            f"RETURN node.{self.text_node_property} AS text, score, "
```

Comment:
we should wrap the property in backticks, to make sure that special chars don't break it.